hasura.planx.uk/migrations/1730277860890_run_sql_migration/up.sql

```diff
@@ -0,0 +1,36 @@
+CREATE OR REPLACE FUNCTION grant_new_user_template_access() RETURNS trigger AS $$
+DECLARE
+    templates_team_id INT;
+    is_demo_user BOOLEAN;
+BEGIN
+    SELECT EXISTS (
+        SELECT 1 
+        FROM team_members 
+        WHERE user_id = NEW.id 
+        AND role = 'demoUser'
+    ) INTO is_demo_user;
+    
+    -- Demo user should not get access as a teamEditor for the templates team...
+    IF is_demo_user THEN
+        RETURN NULL;
+    END IF;
+    
+    -- ...but all other users should
+    SELECT id INTO templates_team_id FROM teams WHERE slug = 'templates';
+    IF templates_team_id IS NOT NULL THEN
+        INSERT INTO team_members (user_id, team_id, role) 
+        VALUES (NEW.id, templates_team_id, 'teamEditor');
+    END IF;
+    
+    RETURN NULL;
+END;
+$$ LANGUAGE plpgsql;
+
+DROP TRIGGER IF EXISTS grant_new_user_template_team_access ON users;
+
+CREATE CONSTRAINT TRIGGER grant_new_user_template_team_access
+AFTER INSERT ON users 
+DEFERRABLE INITIALLY DEFERRED 
```

Comment:
`DEFERRABLE INITIALLY DEFERRED` is new to me!

This ensures that the the trigger will run post _commit_, not post _insert_.

The insert of a new user and the associated `demoUser` role are grouped into a single transaction via [the GraphQL mutation here](https://github.com/theopensystemslab/planx-new/blob/main/editor.planx.uk/src/pages/FlowEditor/components/Team/queries/createAndAddUserToTeam.tsx#L31). 

Without `DEFERRABLE INITIALLY DEFERRED`, the trigger will run after the user has been inserted and then hit a race condition against insertion of the `team_member` record (it may check for the record before it's been inserted).

This will now wait until both the `users` record and `team_members` record has been inserted before running.